firedrake/bcs.py

```diff
@@ -257,32 +423,169 @@ def apply(self, r, u=None):
+                V = eq.lhs.arguments()[0].function_space()
+            bcs = solving._extract_bcs(bcs)
+            # Jp_eq_J and is_linear are evaluated as the tree is constructed
+            self.Jp_eq_J = Jp is None and all([bc.Jp_eq_J for bc in bcs])
+            self.is_linear = all(bc.is_linear for bc in bcs)
+
+            from firedrake.variational_solver import check_pde_args
+            # linear
+            if isinstance(eq.lhs, ufl.Form) and isinstance(eq.rhs, ufl.Form):
+                J = eq.lhs
+                Jp = Jp or J
+                if eq.rhs == 0:
+                    F = ufl_expr.action(J, u)
+                else:
+                    if not isinstance(eq.rhs, (ufl.Form, slate.slate.TensorBase)):
+                        raise TypeError("Provided BC RHS is a '%s', not a Form or Slate Tensor" % type(eq.rhs).__name__)
+                    if len(eq.rhs.arguments()) != 1:
+                        raise ValueError("Provided BC RHS is not a linear form")
+                    F = ufl_expr.action(J, u) - eq.rhs
+            # nonlinear
+            else:
+                if eq.rhs != 0:
+                    raise TypeError("RHS of a nonlinear form equation has to be 0")
+                F = eq.lhs
+                J = J or ufl_expr.derivative(F, u)
+                Jp = Jp or J
+                # Argument checking
+                check_pde_args(F, J, Jp)
+                self.is_linear = False
+            # EquationBCSplit objects for `F`, `J`, and `Jp`
+            self._F = EquationBCSplit(F, u, sub_domain, bcs=[bc if isinstance(bc, DirichletBC) else bc._F for bc in bcs], method=method, V=V)
+            self._J = EquationBCSplit(J, u, sub_domain, bcs=[bc if isinstance(bc, DirichletBC) else bc._J for bc in bcs], method=method, V=V)
+            self._Jp = EquationBCSplit(Jp, u, sub_domain, bcs=[bc if isinstance(bc, DirichletBC) else bc._Jp for bc in bcs], method=method, V=V)
+        elif all(isinstance(args[i], EquationBCSplit) for i in range(3)):
+            # reconstruction for splitting `solving_utils.split`
+            self.Jp_eq_J = Jp_eq_J
+            self.is_linear = is_linear
+            self._F = args[0]
+            self._J = args[1]
+            self._Jp = args[2]
+        else:
+            raise TypeError("Wrong EquationBC arguments")
+
+    def __iter__(self):
+        yield from itertools.chain(self._F)
+
+    def dirichlet_bcs(self):
+        # _F, _J, and _Jp all have the same DirichletBCs
+        yield from self._F.dirichlet_bcs()
+
+    def reconstruct(self, V, subu, u, field):
+        _F = self._F.reconstruct(field=field, V=V, subu=subu, u=u)
+        _J = self._J.reconstruct(field=field, V=V, subu=subu, u=u)
+        _Jp = self._Jp.reconstruct(field=field, V=V, subu=subu, u=u)
+        if all([_F is not None, _J is not None, _Jp is not None]):
+            return EquationBC(_F, _J, _Jp, Jp_eq_J=self.Jp_eq_J, is_linear=self.is_linear)
+
+
+class EquationBCSplit(BCBase):
+
```

Comment:
Although this is only used internally, might be useful to have a docstring attached that describes the function of this object.

Reply:
Added!